<a href="https://colab.research.google.com/github/idoalltime-glitch/data-science-ex1-Ido-Wazana./blob/main/data_science_ex1_Ido_Wazana.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


## 2. Dataset Selection

* **2.1 Project Overview (What, Why, and What We Learn):** In simple terms, we are analyzing how Israeli electricity consumers are switching from the traditional national monopoly to private electricity suppliers. We want to measure how many switching requests succeed, how many get rejected, and what the main reasons for rejection are. This dataset teaches us whether the electricity market reform is working smoothly in practice or where the main friction points are.

* **2.2 Data Source:** Israel Official Government Data Portal (`data.gov.il`).

* **2.3 Collection Purpose:** Tracking, monitoring, and analyzing consumer requests to switch electricity suppliers to private companies as part of the electricity market reform in Israel.

* **2.4 Collecting Entity:** Noga - Israel Independent System Operator Ltd. (ניהול מערכת החשמל) & The Israeli Electricity Authority.

* **2.5 Domain Knowledge Link:** As an Electrical Engineering background context, the Israeli energy sector is undergoing a major structural reform designed to decentralize the electricity market from a historical monopoly into a competitive ecosystem. Consumers can switch from Israel Electric Corporation (IEC) to private providers. Analyzing this dataset provides operational insights into consumer adoption rates, sector preferences (residential vs. commercial), and procedural friction points (e.g., specific rejection reasons for switching).

* **2.6 Personal Context & Domain Knowledge:** I have a strong personal and professional interest in this topic. For the past 4 years, I have been working in the Israeli power sector as a control room operator at the IPM power station—a private natural gas power plant built by Siemens Energy. Having daily experience on the power generation and grid control side, I am excited to analyze the consumer adoption side. Connecting my practical industry experience with this dataset gives a full picture of how Israel's energy reform operates from generation to the end user.

In [20]:
import os
import time
import pandas as pd

file_name = 'electricity_data.xlsx'

# --- 3.1 File Analysis ---
file_size_bytes = os.path.getsize(file_name)
file_size_kb = file_size_bytes / 1024
file_size_mb = file_size_kb / 1024
file_extension = os.path.splitext(file_name)[1]
mod_time = time.ctime(os.path.getmtime(file_name))

print("==========================================")
print("=== Section 3.1: File Analysis Results ===")
print("==========================================")
print(f"File Format: {file_extension}")
print(f"File Size: {file_size_kb:.2f} KB ({file_size_mb:.2f} MB)")
print(f"Last Modified Time: {mod_time}\n")

# --- 3.2 Data Structure ---
df = pd.read_excel(file_name)

print("=============================================")
print("=== Section 3.2: Data Structure Results ===")
print("=============================================")
print(f"Number of Rows: {df.shape[0]}")
print(f"Number of Columns: {df.shape[1]}")
print("\nColumn Names & Data Types:")
print(df.dtypes)

print("\n--- Preview of First 5 Rows ---")
display(df.head())

=== Section 3.1: File Analysis Results ===
File Format: .xlsx
File Size: 114.12 KB (0.11 MB)
Last Modified Time: Sun Jul 26 22:34:07 2026

=== Section 3.2: Data Structure Results ===
Number of Rows: 2638
Number of Columns: 8

Column Names & Data Types:
שנה/ חודש                    datetime64[ns]
סוג אסדרה                            object
תחרות באספקה/ אסדרה קיימת            object
ביתי/ לא ביתי                        object
סיבות לססטוס                         object
פירוט סיבת סטטוס                     object
סטטוס                                object
מספר בקשות                            int64
dtype: object

--- Preview of First 5 Rows ---


,שנה/ חודש,סוג אסדרה,תחרות באספקה/ אסדרה קיימת,ביתי/ לא ביתי,סיבות לססטוס,פירוט סיבת סטטוס,סטטוס,מספר בקשות
0,2026-06-01,מספקים עם אמצעי ייצור,אסדרה קיימת,ביתי,אושר לניוד,תהליך הניוד הושלם,הושלם,30353
1,2026-06-01,מספקים עם אמצעי ייצור,אסדרה קיימת,לא ביתי,אושר לניוד,תהליך הניוד הושלם,הושלם,1558
2,2026-06-01,מספקים וירטואליים,תחרות באספקה,ביתי,אושר לניוד,תהליך הניוד הושלם,הושלם,49152
3,2026-06-01,מספקים וירטואליים,תחרות באספקה,לא ביתי,אושר לניוד,תהליך הניוד הושלם,הושלם,1647
4,2026-06-01,מספקים עם אמצעי ייצור,אסדרה קיימת,ביתי,NaN,לא קיים מונה פעיל בחוזה זה,נדחה,1


## 3. Data Meta-Analysis

### 3.1 File Analysis
* **File Size:** 114.12 KB (0.11 MB).
* **File Format:** Microsoft Excel Spreadsheet (`.xlsx`).
* **System Modification Date:** Sun Jul 26 22:34:07 2026.
* **Data Purpose:** The dataset was generated to maintain regulatory transparency regarding consumer switching requests between electricity providers in Israel, tracking monthly volumes, request statuses, and rejection reasons.

### 3.2 Data Structure
* **Rows and Columns:** The dataset consists of **2,638 rows** and **8 columns**.
* **Column Names & Data Types:**
  * `שנה/ חודש` (*Year / Month*): Temporal Identifier (`datetime64[ns]`)
  * `סוג אסדרה` (*Regulation Type*): Categorical (`object`)
  * `תחרות באספקה/ אסדרה קיימת` (*Supply Competition vs. Existing Framework*): Categorical (`object`)
  * `ביתי/ לא ביתי` (*Residential vs. Non-Residential Sector*): Categorical (`object`)
  * `סיבות לסטטוס` (*Status Category/Reason*): Categorical (`object`)
  * `פירוט סיבת סטטוס` (*Detailed Rejection Reason*): Categorical (`object`)
  * `סטטוס` (*Request Outcome Status - e.g., Completed/Rejected*): Categorical (`object`)
  * `מספר בקשות` (*Number of Requests*): Discrete Numeric Count (`int64`)

* **Discussion & Potential Biases:**
  * **Operational Context:** Data is aggregated monthly for reporting and tracking the efficiency of the electricity reform in Israel.

  * **Aggregation Level:** Unlike row-per-consumer datasets, each record here represents an aggregated group of requests for a specific monthly category.
  
  * **Potential Biases:** Administrative delays in status reporting from private suppliers may cause minor temporal shifts.

> **Note on Column Requirement:** While the raw dataset currently holds 8 columns, feature engineering (extracting Year, Month, Quarter, and Success Rate metrics) will expand the feature set to meet and exceed the 10-column requirement during the data cleaning stage.

In [21]:

missing_data = pd.DataFrame({
    'Missing Count': df.isnull().sum(),
    'Percentage (%)': ((df.isnull().sum() / len(df)) * 100).round(2)
})
missing_data = missing_data[missing_data['Missing Count'] > 0].sort_values(by='Missing Count', ascending=False).reset_index()
missing_data.columns = ['Column Name', 'Missing Count', 'Percentage (%)']
display(missing_data)

,Column Name,Missing Count,Percentage (%)
0,סיבות לססטוס,12,0.45


**4.1 Missing Data**

* **Extent of Missing Data:** Only one column, סיבות לסטטוס (Status Reasons), has missing values—specifically 12 missing rows (0.45%). All other columns are 100% complete.

* **Missingness Pattern (Random vs. Structured):** This is a random pattern (MCAR). Because it is a tiny amount (less than 0.5%), it is likely caused by minor system logging omissions rather than a structural rule.

* **Insights:** The dataset has very high quality (99.55% complete), showing that Noga’s data collection process is reliable.

* **Handling Strategy (Imputation):** We will fill these 12 missing values with the text 'Not Specified' using df['סיבות לסטטוס'].fillna('Not Specified'). This keeps all rows in the dataset without creating errors in our analysis.





In [22]:
# הגדרת העמודות לבדיקה חלקית בצורה בטוחה (ללא עמודת הסיבות)
subset_cols = [col for col in df.columns if 'סיבות' not in col]

# יצירת טבלת הסיכום
dup_summary = pd.DataFrame({
    'Duplicate Type': ['Full Duplicates', 'Partial Duplicates'],
    'Count': [df.duplicated().sum(), df.duplicated(subset=subset_cols).sum()]
})

display(dup_summary)

,Duplicate Type,Count
0,Full Duplicates,0
1,Partial Duplicates,0


**4.2 Duplications**

* **Full & Partial Duplicates:** There are 0 full duplicates and 0 partial duplicates in the dataset (0%).

* **Insights:**
The complete absence of duplicate records shows high data quality and reliable tracking by Noga.

* **Handling Strategy:**
No action is required since there are no duplicate rows to remove.